components/ui/card-collection/use-card-collection-configuration.tsx

```diff
@@ -0,0 +1,179 @@
+
+  newCardStates = [...newCardStates, ...fixedCardStates.map((cardState) => ({ ...cardState, currentColumnCount: cardState.columnCount }))];
+  newCardStates = [...newCardStates, ...expandableCardStates.map((cardState) => ({ ...cardState, currentColumnCount: cardState.isExpanded ? cardState.expandedColumnCount : 0 }))];
+
+  const remainingColumnCount = columnCountToFill - sumBy(newCardStates, (cardState) => cardState.currentColumnCount);
+  const columnsPerRemainingFillRemainingSpaceCardStates = remainingColumnCount / fillRemainingSpaceCardStates.length;
+
+  newCardStates = [ ...newCardStates, ...fillRemainingSpaceCardStates.map((cardState) => ({ ...cardState, currentColumnCount: columnsPerRemainingFillRemainingSpaceCardStates }))];
+
+  return newCardStates;
+};
+
+const getCardStatesWithExpansionStatusChanged = (input: {
+  cardStates: CardStateWithoutCurrentColumnCount[];
+  cardId: string;
+  isExpanded: boolean;
+}): CardStateWithoutCurrentColumnCount[] => {
+  const { cardStates, cardId } = input;
+
+  const elementToChange = cardStates.find(
+    (element) => element.cardId === cardId
+  );
+
+  if (!elementToChange) {
+    return cardStates;
+  }
+
+  if (!isExpandableCardStateWithoutCurrentColumnCount(elementToChange)) {
+    return cardStates;
+  }
+
+  const newElement = { ...elementToChange, isExpanded: input.isExpanded };
+
+  const newCardStates = [ ...cardStates.filter((element) => element.cardId !== cardId), newElement ];
+
+  return newCardStates;
+};
+
+const getCardStatesUpdatedFromCardData = (input: {
+  cardStates: CardStateWithoutCurrentColumnCount[];
+  cardInputData: CardInputData[];
+}): CardStateWithoutCurrentColumnCount[] => {
+  const { cardStates, cardInputData } = input;
+
+  return cardInputData.map((cardInputData) => {
+    if (isFixedCardInputData(cardInputData)) {
+      return {
+        ...cardInputData,
+        currentColumnCount: cardInputData.columnCount,
+      };
+    }
+
+    if (isFillRemainingSpaceCardInputData(cardInputData)) {
+      return {
+        ...cardInputData,
+      };
+    }
+
+    const existingExpandableCardState = cardStates.find(
+      (element) => element.cardId === cardInputData.cardId
+    );
+
+    if (existingExpandableCardState && isExpandableCardInputData(existingExpandableCardState)) {
+      return {
+        ...cardInputData,
+        isExpanded: existingExpandableCardState.isExpanded,
+      };
+    }
+
+    return {
+      ...cardInputData,
+      isExpanded: cardInputData.isExpandedByDefault,
+    };
+  });
+};
+
+interface UseCardCollectionConfigurationInput {
+  cardInputData: CardInputData[];
+  columnCountToFill: number;
+}
+
+interface UseCardCollectionConfigurationOutput {
+  cardConfigurations: CardConfiguration[];
+}
+
+export const useCardCollectionConfiguration = (
+  input: UseCardCollectionConfigurationInput
+): UseCardCollectionConfigurationOutput => {
+  const [cardStates, _setCardStates] = useState<CardState[]>([]);
```

Comment:
The state variable is basically just to store the current column counts based on which parts of the card collection are expanded and minimized.